### PR TITLE
Feat/us 031  Alta de reseña

### PR DESCRIPTION
**Pull request by kuhamaven**
## Description

Creadas las clases de Review, Book tiene su lista de reviews. Hay tests unitarios para todos los criterios de aceptación, y testearlo con Postman tambien funcionó.
La reseña/review permite que un Usuario NO administrador, que haya tenido previamente un prestamo de un libro pueda subirle una reseña al mismo. Esta reseña puede o no tener una descripción, tiene un valor entre 0 y 5.
Los tests unitarios prueban cada uno de estos aspectos
-Un usuario que si tuvo un libro puede subirle una review
-Un admin NO puede subir reviews
-Una review con puntajes invalidos no puede ser subida
-No puedo crear una review si no pedi el libro antes

Cualquier duda respecto a como testearlo en postman estoy a disposición, pero el proceso es el siguiente:
Loggeas un usuario
Tomás un prestamo de un libro (El libro de id 19 funciona ya que tiene copias disponibles)
Una vez recibido el OK del prestamo, podés hacer un POST a http://localhost:8080/review/create/19
-Tenés que ponerle el Authorization Header
-El body debe tener un ReviewModel valido:
{
  "description" : "",
  "value" : "",
  "userModel" : _UserModelValido_,
  "active" : true
}

## ClubHouse Link

https://app.clubhouse.io/bibliotecame/story/123/alta-de-rese%C3%B1a

## Checklist
- [ X ] I've created and run successfully the related unit tests
- [ X ] I've tested the related endpoints with Postman
- [ X ] This pull request has 'sprint_x' as the base branch
- [ X ] This pull request has a descriptive title and description
- [ X ] This PR has no conflicts to merge 
- [ X ] I've added the QA Leader as a reviewer

### CODE REVIEW

File: src/main/java/bibliotecame/back/Review/ReviewController.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<ReviewModel> getReviewModel(@PathVariable Integer id){
+        //Este método será ampliado y modificado por quien corresponda cuando se haga la US "Visualizar reseña"
+        return new ResponseEntity<>(this.reviewService.findReviewById(id), HttpStatus.OK);
+    }
+
+    @PostMapping("/create/{bookId}")
+    public ResponseEntity<ReviewModel> createReview(@Valid @RequestBody ReviewModel reviewModel, @PathVariable Integer bookId){
```
Comment (GustavoCarvallo) [inline src/main/java/bibliotecame/back/Review/ReviewController.java:40]: The endpoint should only receive:
- The book id (on the URL)
- The value of the review (on the body)
- Optionally the description of the review (also in the body)

The user we must take it from the logged user and I don't know what is the `'active'` boolean.

**Review by GustavoCarvallo — CHANGES_REQUESTED** [review]
Please review the comment.

**Comment by kuhamaven** [comment]
Testié con postman además de los test unitarios, ahora con mandarle
{   "value" : 5   }
funciona perfecto, y dentro del service le setea el userModel. Si incluimos tambien un "description" : "__Valor__" se le setea correctamente. Si no se envía descripción alguna, para evitar un nullPointer a futuro, el service le setea de descripción un string vacio "" por si las moscas.

File: src/main/java/bibliotecame/back/Review/ReviewController.java [hunk 2]
```diff
@@ -58,15 +58,8 @@ public ReviewController(ReviewService reviewService, UserService userService, Bo
...
-        List<String> copiesIds = new ArrayList<>();
-        for (int i = 0; i < copies.size() ; i++) {
-            copiesIds.add(copies.get(i).getId());
-        }
-        for (int i = 0; i < loans.size() ; i++) {
-            if(copiesIds.contains(loans.get(i).getCopy().getId())) return true;
-        }
-        return false;
+        List<String> copiesIds = bookService.findBookById(bookId).getCopies().stream().map(CopyModel::getId).collect(Collectors.toList());
+        return loans.stream().map(loanModel -> copiesIds.contains(loanModel.getCopy().getId())).reduce(true, ((aBoolean, aBoolean2) -> aBoolean && aBoolean2));
```
Comment (GustavoCarvallo) [inline src/main/java/bibliotecame/back/Review/ReviewController.java]: This line is not working as expected. 
**What is checking:**
If ALL of the user the loans has a copy that belongs to the book given.

**What should be checking:**
If ANY of the user loans has a copy that belongs to the book given.

**How should be fixed**
Changing the `reduce`, the `initial value` must be false and the operation should be `OR (||)`. But please, check the correct working of the method, do some test with postman (user with more than 1 loan).

**Review by GustavoCarvallo — CHANGES_REQUESTED** [review]
Please review de comment.

Also, the endpoint should check if the user has already made a review of that book, if so the endpoint shouldn't let the user to do another review of the same book.

**Comment by kuhamaven** [comment]
Fixed the method, and added a new one for previouslyReviewed, both following this correction:
"What should be checking:
If ANY of the user loans has a copy that belongs to the book given."

Endpoint tests made:
User with 3 **different** books loaned:
-Could write a review for one of those 3 without problems
-If it attempted to write another one for the same book it got rejected
-Could write a review for a different book
-Could write a review without description

**Review by GustavoCarvallo — APPROVED** [review]
LGTM.
Work as expected, good job! 👍

**Pull request merged**